python/ray/tune/schedulers/pb2_utils.py

```diff
@@ -1,70 +1,104 @@
 from sklearn.metrics import pairwise_distances
 from sklearn.metrics.pairwise import euclidean_distances
 
 
-class TV_SquaredExp(Kern):
+class TV_SquaredExp(Kernel):
     """Time varying squared exponential kernel.
     For more info see the TV-GP-UCB paper:
     http://proceedings.mlr.press/v51/bogunovic16.pdf
     """
 
     def __init__(
-        self, input_dim, variance=1.0, lengthscale=1.0, epsilon=0.0, active_dims=None
+        self,
+        variance=1.0,
+        lengthscale=1.0,
+        epsilon=0.1,
+        variance_bounds=(1e-5, 1e5),
+        lengthscale_bounds=(1e-5, 1e5),
+        epsilon_bounds=(1e-5, 0.5),
     ):
-        super().__init__(input_dim, active_dims, "time_se")
-        self.variance = Param("variance", variance)
-        self.lengthscale = Param("lengthscale", lengthscale)
-        self.epsilon = Param("epsilon", epsilon)
-        self.link_parameters(self.variance, self.lengthscale, self.epsilon)
-
-    def K(self, X, X2):
-        # time must be in the far left column
-        if self.epsilon > 0.5:  # 0.5
-            self.epsilon = 0.5
-        if X2 is None:
-            X2 = np.copy(X)
+        self.variance = variance
+        self.lengthscale = lengthscale
+        self.epsilon = epsilon
+        self.variance_bounds = variance_bounds
+        self.lengthscale_bounds = lengthscale_bounds
+        self.epsilon_bounds = epsilon_bounds
+
+    @property
+    def hyperparameter_variance(self):
+        return Hyperparameter("variance", "numeric", self.variance_bounds)
+
+    @property
+    def hyperparameter_lengthscale(self):
+        return Hyperparameter("lengthscale", "numeric", self.lengthscale_bounds)
+
+    @property
+    def hyperparameter_epsilon(self):
+        return Hyperparameter("epsilon", "numeric", self.epsilon_bounds)
+
+    def __call__(self, X, Y=None, eval_gradient=False):
+        X = np.atleast_2d(X)
+        if Y is None:
+            Y = X
+
+        epsilon = np.clip(self.epsilon, 1e-5, 0.5)
+
+        # Time must be in the first column
         T1 = X[:, 0].reshape(-1, 1)
-        T2 = X2[:, 0].reshape(-1, 1)
+        T2 = Y[:, 0].reshape(-1, 1)
         dists = pairwise_distances(T1, T2, "cityblock")
-        timekernel = (1 - self.epsilon) ** (0.5 * dists)
-
-        X = X[:, 1:]
-        X2 = X2[:, 1:]
+        timekernel = (1 - epsilon) ** (0.5 * dists)
 
-        RBF = self.variance * np.exp(
-            -np.square(euclidean_distances(X, X2)) / self.lengthscale
+        # RBF kernel on remaining features
+        X_spatial = X[:, 1:]
+        Y_spatial = Y[:, 1:]
+        rbf = self.variance * np.exp(
+            -np.square(euclidean_distances(X_spatial, Y_spatial)) / self.lengthscale
         )
 
-        return RBF * timekernel
+        K = rbf * timekernel
 
-    def Kdiag(self, X):
-        return self.variance * np.ones(X.shape[0])
+        if eval_gradient:
+            K_gradient_variance = K
+            dist2 = np.square(euclidean_distances(X_spatial, Y_spatial))
```

Comment:
Not sure why this is no longer divided by lengthscale - see the previous code: `dist2 = np.square(euclidean_distances(X, X2)) / self.lengthscale`

Iiuc with the __call__ in the scikitlearn implementation is similar to K + update_gradients_full in the GPy implementation, but the update_gradients_full portion seems different.

Reply:
Yeah this is because sklearn uses logspace